src/compiler/types.rs

```diff
@@ -0,0 +1,185 @@
+use std::rc::Rc;
+
+pub struct Path(String);
+
+impl Path {
+    pub fn new(string: String) -> Self {
+        Self(string)
+    }
+}
+
+#[derive(Copy, Clone, Debug)]
+pub enum SyntaxKind {
+    Unknown,
+    EndOfFileToken,
+    SemicolonToken,
+    AsteriskToken,
+    EmptyStatement,
+    SourceFile,
+}
+
+#[derive(Debug)]
+pub struct BaseNode {
+    pub kind: SyntaxKind,
+}
+
+pub trait NodeInterface {
+    fn kind(&self) -> SyntaxKind;
+}
+
+#[derive(Debug)]
+pub enum Node {
+    Statement(Statement),
+}
+
+impl NodeInterface for Node {
+    fn kind(&self) -> SyntaxKind {
+        match self {
+            Node::Statement(statement) => statement.kind(),
+        }
+    }
+}
+
+#[derive(Debug)]
+pub struct NodeArray {
+    _nodes: Vec<Node>,
+}
+
+impl NodeArray {
+    pub fn new(nodes: Vec<Node>) -> Self {
+        NodeArray { _nodes: nodes }
+    }
+}
+
+pub enum NodeArrayOrVec {
+    NodeArray(NodeArray),
+    Vec(Vec<Node>),
+}
+
+impl From<NodeArray> for NodeArrayOrVec {
+    fn from(node_array: NodeArray) -> Self {
+        NodeArrayOrVec::NodeArray(node_array)
+    }
+}
+
+impl From<Vec<Node>> for NodeArrayOrVec {
+    fn from(vec: Vec<Node>) -> Self {
+        NodeArrayOrVec::Vec(vec)
+    }
+}
+
+#[derive(Debug)]
+pub enum Statement {
+    EmptyStatement(EmptyStatement),
+}
+
+impl NodeInterface for Statement {
+    fn kind(&self) -> SyntaxKind {
+        match self {
+            Statement::EmptyStatement(empty_statement) => empty_statement.kind(),
+        }
+    }
+}
+
+impl From<Statement> for Node {
+    fn from(statement: Statement) -> Self {
+        Node::Statement(statement)
+    }
+}
+
+#[derive(Debug)]
+pub struct EmptyStatement {
+    pub _node: BaseNode,
+}
+
+impl NodeInterface for EmptyStatement {
+    fn kind(&self) -> SyntaxKind {
+        self._node.kind
+    }
+}
+
+impl From<EmptyStatement> for Statement {
+    fn from(empty_statement: EmptyStatement) -> Self {
+        Statement::EmptyStatement(empty_statement)
+    }
+}
+
+#[derive(Debug)]
+pub struct SourceFile {
+    pub _node: BaseNode,
+    pub statements: NodeArray,
+}
+
+pub trait Program {
+    fn get_source_files(&self) -> &[Rc<SourceFile>];
+    fn get_semantic_diagnostics(&self) -> Vec<Box<dyn Diagnostic>>;
+}
+
+pub enum StructureIsReused {
+    Not,
+    Completely,
+}
+
+pub enum ExitStatus {
+    Success,
+    #[allow(non_camel_case_types)]
+    DiagnosticsPresent_OutputsGenerated,
+}
+
+#[derive(Debug)]
+pub struct ParsedCommandLine {
+    pub file_names: Vec<String>,
+}
+
+pub struct CreateProgramOptions<'config> {
+    pub root_names: &'config [String],
+}
+
+#[non_exhaustive]
+pub struct CharacterCodes;
+#[allow(non_upper_case_globals)]
+impl CharacterCodes {
+    pub const asterisk: char = '*';
+    pub const semicolon: char = ';';
+    pub const slash: char = '/';
+}
```

Comment:
Same question as above somewhere: these seem like they could be public `const`s in a module used like `character_code::asterisk` rather than in a struct.